lib/protocol/ConnectionManager.js

```diff
@@ -13,52 +13,49 @@
 // language governing permissions and limitations under the License.
 'use strict';
 
-var State = require('./ConnectionManagerState');
 var util = require('../util');
 var debug = util.debuglog('hdb_conn');
 
 module.exports = ConnectionManager;
 
 function ConnectionManager(settings) {
-  this.state = new State(settings);
+  this._settings = settings;
 }
 
-ConnectionManager.prototype.updateState = function updateState(newOptions) {
-  this.state.update(newOptions);
-};
-
-ConnectionManager.prototype.getConnectOptions = function getConnectOptions() {
-  return this.state.options.connect;
+ConnectionManager.prototype.updateConnectOptions = function updateConnectOptions(settings) {
+  return util.extend(this._settings, settings);
 };
 
 ConnectionManager.prototype.openConnection = function openConnection(conn, cb) {
-  var dbHosts = this.state.dbHosts;
+  var s = this._settings;
+  var dbHosts = (Array.isArray(s.hosts) && s.hosts.slice()) || [{'host': s.host, 'port': s.port}]
   debug('Opening connection. Hosts:', dbHosts);
 
-  this._openConnectionToHost(conn, dbHosts, 0, [], cb);
-};
+  var errorStats = [];
+  var self = this;
 
-ConnectionManager.prototype._openConnectionToHost = function _openConnectionToHost(conn, dbHosts, whichHost, errorStats, cb) {
-  if (whichHost === dbHosts.length) {
-    return cb(couldNotOpenConnectionError(errorStats));
+  function _tryConnect() {
+    if (dbHosts.length === 0) {
+      return cb(couldNotOpenConnectionError(errorStats));
+    }
+    var settings = util.extend({}, self._settings, dbHosts.shift());
+    self._openConnection(conn, settings, function (err) {
+      if (err) {
+        debug('Connection to %s:%d failed with:', settings.host, settings.port, err.message);
+        errorStats.push({ host: settings.host, port: settings.port, err: err });
+        _tryConnect();
+      } else {
+        debug('Successful connection to %s:%d', settings.host, settings.port);
+        return cb(null);
+      }
+    });
   }
 
-  var self = this;
-  var currentHostOptions = dbHosts[whichHost];
-  var openOptions = util.extend({}, currentHostOptions, this.state.options.encryption);
-  this._openConnection(conn, openOptions, function (err) {
-    if (!err) {
-      debug('Successful connection to %s:%d', openOptions.host, openOptions.port);
-      return cb(null);
-    }
-    debug('Connection to %s:%d failed with:', openOptions.host, openOptions.port, err.message);
-    errorStats.push({ host: openOptions.host, port: openOptions.port, err: err });
-    self._openConnectionToHost(conn, dbHosts, (whichHost + 1), errorStats, cb);
-  });
+  _tryConnect();
 };
 
 ConnectionManager.prototype._openConnection = function _openConnection(conn, openOptions, cb) {
-  var dbName = this.state.options.multiDb.databaseName;
+  var dbName = openOptions.databaseName;
```

Comment:
Better stick with one of the 2 terms - `settings` or `options` and use it all across the file - for variables, methods and members.